pkg/translator/prometheusremotewrite/helper.go

```diff
@@ -175,12 +170,12 @@ func createAttributes(resource pcommon.Resource, attributes pcommon.Map, externa
 		l[name] = extras[i+1]
 	}
 
-	labels = labels[:0]
+	startIndex := len(c.labels)
 	for k, v := range l {
-		labels = append(labels, prompb.Label{Name: k, Value: v})
+		c.labels = append(c.labels, prompb.Label{Name: k, Value: v})
 	}
 
-	return labels
+	return c.labels[startIndex:]
```

Comment:
Is there any possibility that `len(c.labels)` is not 0 here? It's reset every time we call FromMetrics and I couldn't find any other place in the code where we write to this array, so why not just return `c.labels` and not worry about startIndex? I might be missing something but it feels like we're overcomplicating things here

```go
for k, v := range l {
	c.labels = append(c.labels, prompb.Label{Name: k, Value: v})
}
return c.labels
```

Reply:
`startIndex` is important for keeping the returned slices isolated from each other while sharing the same underlying array within a single `FromMetrics` call. It is 0 only for the first series of a batch.

Here is how it works: `FromMetrics` is called once per batch, and `createAttributes` for every series within the batch. We want to re-use the backing array of the `labels` slice for all series within a single batch. We do that by appending the labels of each series to the end of the slice. Finally we return only starting from `startIndex` so the caller doesn't see labels from other series (while reusing the same backing array which naturally grows up to the size needed to fit a single `FromMetrics` call).

For example, if X1...X4 are labels from series X and Y1...Y3 are labels from series Y, then the backing array of `c.labels` will look like `[X1, X2, X3, X4, Y1, Y2, Y3]`. Then on the next `FromMetrics` call the index is reset to 0 and we can re-use the entire array with zero allocations.

Reply:
Perfect, thank you for the explanation :)

Now what I'm thinking is if we have tests that assure concurrency works. Mostly to make sure we don't break the thread-safe promise by accident